addons/hr_payroll/models/hr_payroll.py

```diff
@@ -97,6 +97,130 @@ def get_all_structures(self):
+            :return: Tuple of 3 elements ( date from, date to, type of leave )
+            """
+            res = []
+            for leave in leaves:
+                # The cases here are not mutually exclusive, so the order of the test is important
+                #   First we test if the leave is sub-interval of the given interval
+                #   than we test if the leave is overlapping left or right with the interval
+                if leave[0] >= date_from and leave[1] <= date_to:
+                    # If the leave is sub-iterval of the work interval
+                    res.append(leave)
+                elif leave[0] < date_from and leave[1] > date_to:
+                    # If the work interval is sub-interval of the leave
+                    res.append((date_from, date_to, leave[2]))
+                elif leave[0] <= date_to and leave[1] >= date_to:
+                    if limit:
+                        # The case leave[0] > date_from_to is alrady handled with the first if block, so we limit from the start of period
+                        res.append((leave[0], date_to, leave[2]))
+                    else:
+                        res.append(leave)
+                elif leave[0] <= date_from and leave[1] >= date_from:
+                    if limit:
+                        # The case leave[1] < date_to is alrady handled with the first if block, so we limit to the end of period
+                        res.append((date_from, leave[1], leave[2]))
+                    else:
+                        res.append(leave)
+            return res
+
+        def hours_in_interval(interval):
+            """
+                Compute the duration in hours of the intreval
+            :param interval: Interval tuple (date from, date to)
+            :return: Total hours in the interval
+            """
+            return (interval[1] - interval[0]).total_seconds() / 3600.0
+
+        res = []
+        # Read all leaves for the period
+        leave_ids = self.get_resource_leaves(date_from, date_to)
+        # Prepare the list for get_working_intervals_of_day
+        leave_intervals = map(lambda x: (
+        fields.Datetime.from_string(x.date_from), fields.Datetime.from_string(x.date_to), x.holiday_status_id.name),
+                              leave_ids)
+
+        day_from = fields.Datetime.from_string(date_from)
+        day_to = fields.Datetime.from_string(date_to)
+        nb_of_days = (day_to - day_from).days + 1
+
+        # Gather all intervals and holidays
+        for day in range(0, nb_of_days):
+            working_intervals_on_day = self.working_hours.get_working_intervals_of_day(
+                start_dt=day_from + timedelta(days=day))
+            total_working_hours_of_day = sum(map(lambda x: hours_in_interval(x), working_intervals_on_day))
+
+            for interval in working_intervals_on_day:
+                intervals_removed_leaves = self.working_hours.interval_remove_leaves(interval, leave_intervals)
```

Comment:
We are cleaning the schedule intervals from the leaves in the interval so we will have the real working hours.